functions/src/winloss.ts

```diff
@@ -49,19 +49,17 @@ export async function endGameProcess(
   transaction: firestore.Transaction
 ) {
   // apply the stats
-  let { players } = lobbyData;
-  const { uids, winner } = lobbyData;
+  const { players, winner } = lobbyData;
 
   // update each players doc
   await Promise.all(
-    uids.map(async (uid, index) => {
+    Object.entries(lobbyData.players).map(async ([uid, player]) => {
       const userDocRef = userCollection.doc(uid);
       const userDoc = await transaction.get(userDocRef);
 
-      if (players[index].role == undefined) {
-        functions.logger.error("This player's role doesn't exist!");
+      if (player.role == undefined) {
         return;
-      } else if (players[index].role == "SPECTATOR") {
+      } else if (players[uid].role == "SPECTATOR") {
```

Comment:
why not just use `player`?

Reply:
I might not be understanding what you mean but putting "player" in place of "uid" gives me "Type 'Player' cannot be used as an index type". Above I am using object.entries and getting the uid and player map as the key and value. I just assumed I had to use uid since that is the key we want

Reply:
I mean `player` instead of `players[uid]`

Reply:
ohhh gotcha okay